### ISSUE TITLE
Variables for CLI or tfvars not being passed to modules

### ISSUE THREAD

**Issue by mikemcguire** [hashicorp/terraform#15458]
Is there something wrong with passing variables to modules or am I completely reading the documentation wrong?
I've got a project I've broken into the following directory structure.
```
    master.tf
    variables.tfvars
    - providers/
    -- digital_ocean/
    --- digital_ocean.tf
    --- variables.tf
    -- cloud_flare/
    --- cloud_flare.tf
    --- variables.tg
    - management/
    -- jenkins/
    --- jenkins-master.tf
```
I'm attempting to pass my Digital Ocean and Cloudflare tokens as variables, to their respective modules. Everything below the root directory is loaded into `master.tf` as a module.

I have the following in my varaibles.tfvars file:
```
    cloudflare_email  ="service@email.com"
    cloudflare_token  ="TOKEN_STRING"
    do_token          ="${DO_PAT}"
```
The following lines appear in my `master.tf`
```
    variable "do_token" {}
    module "digital_ocean" {
        source          = "./providers/digital_ocean"
        token           = "${var.do_token}"
    }
    
    
    variable "cloudflare_email" {}
    variable "cloudflare_token" {}
    module "cloud_flare" {
        source          = "./providers/cloud_flare"
        email = "${var.cloudflare_email}"
        token = "${var.cloudflare_token}"
    }
```
My digital_ocean module looks like 
```
    variable "token" {}
    
    provider "digitalocean" {
      token = "${var.token}"
    }
```
and the cloudflare provider looks like
```
    variable "email" {}
    variable "token" {}
    
    provider "CloudFlare" {
        email = "${var.email}"
        token = "${var.token}"
    }
```

Setting up my jenkins master server on DO
```
    resource "digitalocean_droplet" "jenkins-master" {
    ...
    }
```
From the command line I'm running `terraform apply -var-file="variables.tfvars"`

or I've also tried passing them via the CLI like so..
```
    terraform apply \
      -var "cloudflare_email=service@email.com" \
      -var "cloudflare_token=TOKEN_STRING" \
      -var "do_token=${DO_PAT}"
```
With the above declarations, it will send me into UI mode and prompt me for those variables rather than reading them automatically. I've replicated this behavior on both Terraform v0.9.8 and v0.9.10.

Before I started breaking everything out into separate modules, passing in variables presented no issues. 

I've tried pulling the provider declarations into `master.tf` to see if there were any weird behaviors with modularizing them, with the same behavior. 

I also tried hard coding the values into the provider declarations and am experiencing the same behaviors.

**Comment by rorychatt** [comment]
Hi Mike,

Two issues here :)

1. **.tfvars** files must be named **terraform**.tfvars to function correctly.

2.  As for _provider.digitalocean.token_ , the issue isn't actually related to variables - rather it is that Terraform is calling the Jenkins Module before it calls the Digital Ocean provider module. As no provider has been set at the time it fails. 

To be honest - it's pretty uncommon to call Providers in a module, and to be honest - not all that necessary as you are actually increasing LOC (3 lines to call the provider, 4 to call the module).

You have a few options:

- **Set the provider details at the top level function**
- **Set the Provider details at an environment level:** This will make your life easier when you start triggering Terraform from providers like Jenkins, and adheres to 12 Factor App.

**Comment by mikemcguire** [comment]
It actually runs fine if you provide the `-var-file` flag. Per my original comment I'm running `terraform apply -var-file="variables.tfvars"`

I was actually curious to use it as a module as it would be ideal to have multiple providers for my use case.  As we have separated accounts for billing/clients. Granted the DO provider declaration is really simple,  if I wanted to write some custom providers for vendors that currently don't have a provider, it would be ideal to keep the pattern consistent. 

Not a big deal though I've discussed this in detail on stack overflow and found a few other times this issue has been opened  and closed by the terraform team.. Just declaring the provider directly for now.
